packages/core/src/data_sources/model/ComponentWithCollectionsState.ts

```diff
@@ -0,0 +1,119 @@
+import { DataCollectionStateMap } from '../../data_sources/model/data_collection/types';
+import DataResolverListener from '../../data_sources/model/DataResolverListener';
+import DataVariable, { DataVariableProps, DataVariableType } from '../../data_sources/model/DataVariable';
+import Components from '../../dom_components/model/Components';
+import Component from '../../dom_components/model/Component';
+import { ObjectAny } from '../../common';
+import DataSource from './DataSource';
+import { isArray } from 'underscore';
+
+export type DataVariableMap = Record<string, DataVariableProps>;
+
+export type DataSourceRecords = DataVariableProps[] | DataVariableMap;
+
+export default class ComponentWithCollectionsState<DataResolverType> extends Component {
+  collectionsStateMap: DataCollectionStateMap = {};
+  dataSourceWatcher?: DataResolverListener;
+
+  constructor(props: any, opt: any) {
+    super(props, opt);
+    this.listenToPropsChange();
+  }
+
+  onCollectionsStateMapUpdate(collectionsStateMap: DataCollectionStateMap) {
+    this.collectionsStateMap = collectionsStateMap;
+    this.dataResolverWatchers?.onCollectionsStateMapUpdate?.();
+
+    this.components().forEach((cmp) => {
+      cmp.onCollectionsStateMapUpdate?.(collectionsStateMap);
+    });
+  }
+
+  syncOnComponentChange(model: Component, collection: Components, opts: any) {
+    const prev = this.collectionsStateMap;
+    this.collectionsStateMap = {};
+    super.syncOnComponentChange(model, collection, opts);
+    this.collectionsStateMap = prev;
+    this.onCollectionsStateMapUpdate(prev);
+  }
+
+  setDataResolver(dataResolver: DataResolverType | undefined) {
+    return this.set('dataResolver', dataResolver);
+  }
+
+  getDataResolver(): DataResolverType | undefined {
+    return this.get('dataResolver');
+  }
+
+  protected listenToDataSource() {
+    const path = this.dataSourcePath;
+    if (!path) return;
+
+    const { em, collectionsStateMap } = this;
+    this.dataSourceWatcher = new DataResolverListener({
+      em,
+      resolver: new DataVariable({ type: DataVariableType, path }, { em, collectionsStateMap }),
+      onUpdate: () => this.onDataSourceChange(),
+    });
+  }
+
+  protected listenToPropsChange() {
+    this.on(`change:dataResolver`, () => {
+      this.listenToDataSource();
+    });
+
+    this.listenToDataSource();
+  }
+
+  protected get dataSourceProps(): DataVariableProps | undefined {
+    return this.get('dataResolver');
+  }
```

Comment:
above you have `getDataResolver()`, maybe let's replace this one with `get dataResolverProps()`?!